sys/ztimer64/ztimer64.c

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (C) 2021 Kaspar Schleiser <kaspar@schleiser.de>
+ *               2021 Freie Universität Berlin
+ *               2021 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     sys_ztimer
+ * @{
+ *
+ * @file
+ * @brief       ztimer64 core functionality
+ *
+ * This file contains ztimer64's main API implementation and functionality
+ *
+ * TODO:
+ * - add some kind of tick so base_clock overflows don't get lost
+ * - TEST
+ * - add proper unittests
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+#include <assert.h>
+#include <stdint.h>
+#include <inttypes.h>
+
+#include "ztimer64.h"
+
+#define ZTIMER64_CHECKPOINT_INTERVAL (1LU << 31)
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static int _add_entry_to_list(ztimer64_clock_t *clock, ztimer64_base_t *entry);
+static int _del_entry_from_list(ztimer64_clock_t *clock,
+                                ztimer64_base_t *entry);
+static void _ztimer64_update(ztimer64_clock_t *clock);
+
+/* debug aid */
+void ztimer64_clock_print(const ztimer64_clock_t *clock);
+
+static unsigned _is_set(const ztimer64_t *t)
+{
+    /* no timer can ever trigger at `target==0`, so we use that
+     * as marker */
+    return !!(t->base.target);
+}
+
+unsigned ztimer64_is_set(const ztimer64_t *timer)
+{
+    return _is_set(timer);
+}
+
+void ztimer64_remove(ztimer64_clock_t *clock, ztimer64_t *timer)
+{
+    unsigned state = irq_disable();
+
+    if (_is_set(timer)) {
+        if (_del_entry_from_list(clock, &timer->base) == 0) {
+            _ztimer64_update(clock);
+        }
+    }
+
+    irq_restore(state);
+}
+
+void ztimer64_set_at(ztimer64_clock_t *clock, ztimer64_t *timer, uint64_t val)
+{
+    unsigned state = irq_disable();
+
+    if (_is_set(timer)) {
+        _del_entry_from_list(clock, &timer->base);
+    }
+
+    timer->base.target = val;
+
+    /* optionally subtract a configurable adjustment value */
+    /* there is no API to set this ATM */
+    if (val > clock->adjust_set) {
+        val -= clock->adjust_set;
+    }
+    else {
+        val = 0;
+    }
+
+    /* NOTE: if val is 0 at this point, this will be fixed to "now()" in _ztimer64_update() */
```

Comment:
Do you mean `timer->base.target`?? I dont see how the local variable `val` passes to `ztimer54_udate`

Reply:
well spotted, `timer->base.target` needs to be set *after* applying the adjust. Fixed.